pkg/tbtc/signing.go

```diff
@@ -0,0 +1,224 @@
+package tbtc
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"github.com/keep-network/keep-core/pkg/chain"
+	"github.com/keep-network/keep-core/pkg/protocol/group"
+	"github.com/keep-network/keep-core/pkg/tecdsa/retry"
+	"github.com/keep-network/keep-core/pkg/tecdsa/signing"
+	"golang.org/x/exp/slices"
+	"math/big"
+	"math/rand"
+	"sort"
+)
+
+// signingRetryLoop is a struct that encapsulates the signing retry logic.
+type signingRetryLoop struct {
+	signingGroupMemberIndex group.MemberIndex
+	signingGroupOperators   chain.Addresses
+
+	chainConfig *ChainConfig
+
+	attemptCounter    uint
+	attemptStartBlock uint64
+	attemptSeed       int64
+
+	randomRetryCounter uint
```

Comment:
Why do we need `attemptCounter` and `randomRetryCounter`, can't we just use `attemptCounter` everywhere? I see `randomRetryCounter` is used in `qualifiedOperatorsSet` but I think we could use `attemptCounter` there as well.

Reply:
The main reason is that `attemptCounter` counts from `1` and `randomRetryCounter` from `0` to match the retry algorithm expectations. However, looks like I misread the docs as this is true for keygen retry and basically doesn't matter for the signing retry logic. I'll change to use `attemptCounter` everywhere.

Reply:
See https://github.com/keep-network/keep-core/pull/3260/commits/aabe50e8f6a50c092282be5c31ccff0de6e0d61e